python_onnxifi/python_onnxifi.cc

```diff
@@ -0,0 +1,355 @@
+  // pointer
+  // Otherwise, not present
+  std::unordered_map<onnxBackendID, onnxBackend> initialized_;
+};
+
+const std::unordered_map<onnxEnum, std::string>&
+DeviceIDs::deviceType_to_string_() {
+  static std::unordered_map<onnxEnum, std::string> deviceType_to_string_ = {
+      {ONNXIFI_DEVICE_TYPE_NPU, "NPU"},
+      {ONNXIFI_DEVICE_TYPE_DSP, "DSP"},
+      {ONNXIFI_DEVICE_TYPE_GPU, "GPU"},
+      {ONNXIFI_DEVICE_TYPE_CPU, "CPU"},
+      {ONNXIFI_DEVICE_TYPE_FPGA, "FPGA"},
+      {ONNXIFI_DEVICE_TYPE_HETEROGENEOUS, "HETEROGENEOUS"}};
+  return deviceType_to_string_;
+}
+
+const std::unordered_map<std::string, onnxEnum>&
+DeviceIDs::string_to_deviceType_() {
+  static std::unordered_map<std::string, onnxEnum> string_to_deviceType_ = {
+      {"NPU", ONNXIFI_DEVICE_TYPE_NPU},
+      {"DSP", ONNXIFI_DEVICE_TYPE_DSP},
+      {"GPU", ONNXIFI_DEVICE_TYPE_GPU},
+      {"CPU", ONNXIFI_DEVICE_TYPE_CPU},
+      {"FPGA", ONNXIFI_DEVICE_TYPE_FPGA},
+      {"HETEROGENEOUS", ONNXIFI_DEVICE_TYPE_HETEROGENEOUS}};
+
+  return string_to_deviceType_;
+}
+
+struct BackendRep {
+ public:
+  //ONNXIFI graph that has been initalized
+  //takes ownership of serializedModel used to initialize G
+  BackendRep(std::string&& serializedModel, onnxBackend backend) : graph_(), backend_(backend), serialized_model_(serializedModel), conversion_() {
+    if (onnxInitGraph(backend, serialized_model_.size(), serialized_model_.c_str(), 0, nullptr, &graph_)
+                                         != ONNXIFI_STATUS_SUCCESS) {
+      std::cerr << "Could not initialize graph on given device" << std::endl;
+      graph_ = NULL; //Make dummy BackendRep with NULL graph TODO: Return python None object
+    }
+  }
+
+  // TODO: Implement run
+  //TODO: Change usage of memoryFence (onnxEvent instead of onnxEvent*) when submodule pulled
+  py::dict run(py::dict inputs, py::kwargs kwargs) {
+    conversion_.updateInputDescriptorsData(inputs);
+    auto inputDescriptors = conversion_.getInputTensorDescriptors();
+    ModelProto model;
+    if (!model.ParseFromString(serialized_model_))  {
+      std::cerr << "Failed to parse model proto" << std::endl;
+      //TODO: Return None
+    }
+    conversion_.updateOutputDescriptorsData(model);
+    auto outputDescriptors = conversion_.getOutputTensorDescriptors();
+    if (onnxSetGraphIO(graph_, inputDescriptors.size(), inputDescriptors.data(),
+                               outputDescriptors.size(), outputDescriptors.data()) != ONNXIFI_STATUS_SUCCESS)  {
+      std::cerr << "I/O could not be set" << std::endl;
+      //TODO: Return None
+    }
+    onnxMemoryFence inputFence;
+    inputFence.type = ONNXIFI_SYNCHRONIZATION_EVENT;
+    onnxEvent inputEvent;
+    if (onnxInitEvent(backend_, &inputEvent) != ONNXIFI_STATUS_SUCCESS)  {
+      throw std::runtime_error("Internal Error: Event for input memory fence could not be initialized (expected ONNXIFI_STATUS_SUCCESS)");
```

Comment:
no clang-format?

Reply:
i'm going to put it in a script in a future PR so I don't keep forgetting